crates/quickjs-wasm-rs/src/js_value/callback_arg.rs

```diff
@@ -0,0 +1,233 @@
+use std::{collections::HashMap, convert::TryInto, fmt};
+
+use anyhow::Result;
+
+use super::{qjs_convert::from_qjs_value, JSValue};
+use crate::js_binding::value::JSValueRef;
+
+/// `CallbackArg` is given  to callback functions and can then be lazily evaluated to a `JSValue`.
+///
+/// # Example
+/// ```
+/// let context = JSContextRef::default();
+/// context.wrap_callback(|_ctx, _this, args| {
+///     // Convert args[0] to a Rust `String` type
+///     let s: String = args[0].try_into()?;
+///     println!("{}", s);
+///     Ok(JSValue::Undefined)
+/// })?;
+/// ```
+#[derive(Copy, Clone)]
+pub struct CallbackArg {
+    inner: JSValueRef,
+}
+
+impl CallbackArg {
+    /// Create a new `CallbackArg` with a `JSValueRef`.
+    pub fn new(inner: JSValueRef) -> Self {
+        Self { inner }
+    }
+
+    /// Get the underlying `JSValueRef` value wrapped by this `CallbackArg`. This is used as an escape hatch to operate directly with the raw pointer to the QuickJS value.
+    ///
+    /// # Safety
+    ///
+    /// This function is marked unsafe because `JSValueRef` contains a raw pointer to the underlying QuickJS value so it is possible to cause undefined behavior if the pointer is used incorrectly.
+    pub unsafe fn inner_value(&self) -> JSValueRef {
+        self.inner
+    }
+
+    /// Convert the underlying `JSValueRef` to a Rust `JSValue` type.
+    fn to_js_value(self) -> Result<JSValue> {
+        from_qjs_value(&self.inner)
+    }
+}
+
+impl fmt::Display for CallbackArg {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.to_js_value().unwrap())
+    }
+}
+
+/// A macro to implement the `TryFrom<&CallbackArg>` and `TryFrom<CallbackArg>` trait
+/// for various Rust types.
+macro_rules! try_from_impl {
+    ($($t:ty),+ $(,)?) => {
+        $(impl TryFrom<&CallbackArg> for $t {
+            type Error = anyhow::Error;
+
+            fn try_from(value: &CallbackArg) -> Result<Self> {
+                value.to_js_value()?.try_into()
+            }
+        }
+
+        impl TryFrom<CallbackArg> for $t {
+            type Error = anyhow::Error;
+
+            fn try_from(value: CallbackArg) -> Result<Self> {
+                value.to_js_value()?.try_into()
+            }
+        })+
+    };
+}
+
+try_from_impl!(
+    bool,
+    i32,
+    usize,
+    f64,
+    String,
+    Vec<JSValue>,
+    Vec<u8>,
+    HashMap<String, JSValue>,
+);
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::js_binding::context::JSContextRef;
+
+    #[test]
+    fn test_bool() -> Result<()> {
+        let context = JSContextRef::default();
+        let val = context.eval_global("test.js", "true").unwrap();
```

Comment:
Nit: You can use `?` instead of `.unwrap()` here and if `eval_global` returns an error, it'll fail the test. I find that syntax a bit cleaner since it's one character instead of nine with the same semantics and outcome.